course/page/choice.py

```diff
@@ -65,20 +64,78 @@ def markup_to_html_plain(page_context, s):
     return s
 
 
-# {{{ choice question base
+# {{{ choice data model
 
-class ChoiceQuestionBase(PageBaseWithTitle, PageBaseWithValue):
+class ChoiceModes:
+    INCORRECT = "incorrect"
+    CORRECT = "correct"
+    DISREGARD = "disregard"
+    ALWAYS_CORRECT = "always_correct"
+
+    values = [INCORRECT, CORRECT, DISREGARD, ALWAYS_CORRECT]
+
+
+class ChoiceInfo(object):
     CORRECT_TAG = "~CORRECT~"
     DISREGARD_TAG = "~DISREGARD~"
     ALWAYS_CORRECT_TAG = "~ALWAYS_CORRECT~"
 
+    def __init__(self, mode, text):
+        assert mode in ChoiceModes.values
+
+        self.mode = mode
+        self.text = text
+
+    @classmethod
+    def parse_from_yaml(cls, vctx, location, node):
+        # could be a number or a bool due to sloppy YAML
+        try:
+            node = str(node)
+        except Exception:
+            raise ValidationError(
+                    _("%(location)s: unable to convert to string")
+                    % {'location': location})
+
+        s = node
+
+        item_mode = [None]
+
+        def mode_from_prefix(mode, prefix, s):
+            if s.startswith(prefix):
+                s = s[len(prefix):].strip()
+
+                if item_mode[0] is not None:
+                    raise ValidationError(
+                            _("%(location)s: more than one choice mode set")
+                            % {'location': location})
+
+                item_mode[0] = mode
+
+            return s
+
+        s = mode_from_prefix(ChoiceModes.CORRECT, cls.CORRECT_TAG, s)
+        s = mode_from_prefix(ChoiceModes.DISREGARD, cls.DISREGARD_TAG, s)
+        s = mode_from_prefix(ChoiceModes.ALWAYS_CORRECT, cls.ALWAYS_CORRECT_TAG, s)
```

Comment:
This won't handle `~CORRECT~~CORRECT~foobar`, i.e., it will view `~CORRECT~foobar` as a correct answer.

Reply:
Should it?

Reply:
(sorry for late reponse) `~CORRECT~~DISREGARD~foobar` will raise `ValidationError`, am I right?

Reply:
And `~DISREGARD~~CORRECT~foobar` is also expected to raise a `ValidationError` but not.

Reply:
Agreed,